File: minCoins.go
package main

import "fmt"

func minCoin(coins []int, n int) (int, []int) {
  M := make([]int, n + 1)      //Stores minimum number of coins needed for value i
  C := make([][]int, n + 1)    //Stores which coins are used
  M[0] = 0
  C[0] = []int{}
  for i := 1; i <= n; i++ {
    minVal := int(int(^uint(0) >> 1))
    var minCoins []int
    for _, coin := range coins {
      if i - coin < 0 {
        continue
      }

      localMin := 1 + M[i - coin]
      if minVal > localMin {
        minVal = localMin
        minCoins = append(C[i - coin], coin)
      }
    }
    M[i] = minVal
    C[i] = minCoins
  }
  return M[n], C[n]
}

func main() {
  fmt.Println(minCoin([]int{1, 5, 25, 30}, 37))
}

File: minCoins.java
class minCoins {
  public static void main(String[] args) {
    int coins[] = new int[]{1, 5, 25, 30};
    int change = 37;
    System.out.println(minCoins(coins, change));
  }

  public static int minCoins(int[] coins, int n) {
    int M[] = new int[n + 1];
    M[0] = 0;
    for (int i = 1; i <= n; i++) {
      int minVal = Integer.MAX_VALUE;
      for (int coin: coins) {
        if (i - coin < 0)
          continue;
        minVal = (minVal < 1 + M[i - coin]) ? minVal : (1 + M[i - coin]);
      }
      M[i] = minVal;
    }

    return M[n];
  }
}
